tests/php/ORM/DBReplicaTest.php

```diff
@@ -0,0 +1,230 @@
+            $this->assertSame('replica_01', $this->getLastConnectionName(), "method is $method");
+            (new DataQuery(Group::class))->$method();
+            $this->assertSame('primary', $this->getLastConnectionName(), "method is $method");
+        }
+        // Assert that DataQuery methods with a param respect DataObject.must_use_primary_db
+        $methods = [
+            'max',
+            'min',
+            'avg',
+            'sum',
+            'column',
+        ];
+        foreach ($methods as $method) {
+            (new DataQuery(TestObject::class))->$method('ID');
+            $this->assertSame('replica_01', $this->getLastConnectionName(), "method is $method");
+            (new DataQuery(Group::class))->$method('ID');
+            $this->assertSame('primary', $this->getLastConnectionName(), "method is $method");
+        }
+    }
+
+    public function testRoutes()
+    {
+        // Create a custom rule to test our controller that should default to using a replica
+        $rules = Config::inst()->get(Director::class, 'rules');
+        $rules['test'] = TestController::class;
+        // Ensure that routes staring with '$' are at the bottom of the assoc array index and don't override
+        // our new 'test' route
+        uksort($rules, fn($a, $b) => str_starts_with($a, '$') ? 1 : (str_starts_with($b, '$') ? -1 : 0));
+        // Assert that normal routes use replica
+        $this->get('/test');
+        $this->assertSame('replica_01', $this->getLastConnectionName());
+        // Assert that Security route is forced to use primary
+        $this->get('/Security/login');
+        $this->assertSame('primary', $this->getLastConnectionName());
+        // Assert that dev route is forced to use primary
+        $this->get('/dev/tasks');
+        $this->assertSame('primary', $this->getLastConnectionName());
+    }
+
+    public function testSetCanUseReplicas()
+    {
+        DB::setCanUseReplicas(false);
+        TestObject::get()->count();
+        $this->assertSame('primary', $this->getLastConnectionName());
+        DB::setCanUseReplicas(true);
+        TestObject::get()->count();
+        $this->assertSame('replica_01', $this->getLastConnectionName());
+    }
+
+    public function testConfigMethods()
+    {
+        // Test DB::hasConfig()
+        $this->assertFalse(DB::hasConfig('lorem'));
+        DB::setConfig(['type' => 'lorem'], 'lorem');
+        $this->assertTrue(DB::hasConfig('lorem'));
+        // Test DB::hasReplicaConfig()
+        $this->assertTrue(DB::hasReplicaConfig());
+        $default = DB::getConfig('default');
+        (new ReflectionClass(DB::class))->setStaticPropertyValue('configs', ['default' => $default]);
+        $this->assertFalse(DB::hasReplicaConfig());
+        // Test DB::getReplicaConfigKey()
+        $this->assertSame('replica_03', DB::getReplicaConfigKey(3));
+        $this->assertSame('replica_58', DB::getReplicaConfigKey(58));
+    }
```

Comment:
Are these really heavy enough tests to be worth testing all together like this? Seems like these should be separated.

Reply:
They're not turns out, not sure why it was so slow when I initial thought this. Have split out the tests and using dataProviders where appropriate